File: app/src/main/java/sk/romanstrazanec/particleflowkotlin/javacode/AttractionPoint.java
package sk.romanstrazanec.particleflowkotlin.javacode;

class AttractionPoint {
    float x, y;

    AttractionPoint(float x, float y) {
        this.x = x;
        this.y = y;
    }

    void moveTo(float x, float y) {
        this.x = x;
        this.y = y;
    }

    void attract(Particle p) {
        p.x += Float.compare(x, p.x);
        p.y += Float.compare(y, p.y);
    }
}

File: app/src/main/java/sk/romanstrazanec/particleflowkotlin/AttractionPoint.kt
package sk.romanstrazanec.particleflowkotlin

class AttractionPoint(var x: Float, var y: Float, private val attractionSpeed: Float) {
    fun moveTo(x: Float, y: Float) {
        this.x = x
        this.y = y
    }

    fun attract(p: Particle) {
        val rx = x - p.x
        val ry = y - p.y
        val distance = rx * rx + ry * ry

        if (distance < 0.5f) p.dispersed = true

        val force = (if (p.dispersed) -1f else 1f) * attractionSpeed / distance
        p.moveTo(p.x + force * rx, p.y + force * ry)

        if (rx * p.dx < 0f && ry * p.dy < 0f) p.dispersed = true
    }
}
